src/index.js

```diff
@@ -29,361 +25,450 @@ class IPLDResolver {
+          } catch (error) {
+            return reject(error)
           }
-          return e.basePath
-        }),
-        pull.filter(Boolean)
-      )
-    }
+          treePaths.push(...paths)
+        }
+        const treePath = treePaths.shift()
+        let fullPath = basePath + treePath
+
+        // Only follow links if recursion is intended
+        if (options.recursive) {
+          cid = await maybeRecurse(block, treePath)
+          if (cid !== null) {
+            queue.push({ cid, basePath: fullPath + '/' })
+          }
+        }
 
-    // filter out by path
-    if (path) {
-      return pull(
-        p,
-        pull.map((el) => {
-          if (el.indexOf(path) === 0) {
-            el = el.slice(path.length + 1)
-            return el
+        // Return it if it matches the given offset path, but is not the
+        // offset path itself
+        if (fullPath.startsWith(offsetPath) &&
+            fullPath.length > offsetPath.length) {
+          if (offsetPath.length > 0) {
+            fullPath = fullPath.slice(offsetPath.length + 1)
           }
-        }),
-        pull.filter(Boolean)
-      )
+          return resolve({ done: false, value: fullPath })
+        } else { // Else move on to the next iteration before returning
+          return resolve(next())
+        }
+      })
     }
 
-    return p
-  }
-
-  remove (cids, callback) {
-    this.bs.delete(cids, callback)
+    return fancyIterator({ next })
   }
 
   /*           */
   /* internals */
   /*           */
+  async _getFormat (codec) {
+    // TODO vmx 2019-01-24: Once all CIDs support accessing the codec code
+    // instead of the name, remove this part
+    if (typeof codec === 'string') {
+      const constantName = codec.toUpperCase().replace(/-/g, '_')
+      codec = multicodec[constantName]
+    }
 
-  _get (cid, callback) {
-    waterfall([
-      (cb) => this._getFormat(cid.codec, cb),
-      (format, cb) => this.bs.get(cid, (err, block) => {
-        if (err) return cb(err)
-        cb(null, format, block)
-      }),
-      (format, block, cb) => {
-        format.util.deserialize(block.data, (err, deserialized) => {
-          if (err) {
-            return cb(err)
-          }
-          cb(null, deserialized)
-        })
-      }
-    ], callback)
-  }
-
-  _getFormat (codec, callback) {
     if (this.resolvers[codec]) {
-      return callback(null, this.resolvers[codec])
+      return this.resolvers[codec]
     }
 
     // If not supported, attempt to dynamically load this format
-    this.support.load(codec, (err, format) => {
-      if (err) return callback(err)
-      this.resolvers[codec] = format
-      callback(null, format)
-    })
+    const format = await this.loadFormat(codec)
+    this.resolvers[codec] = format
+    return format
   }
 
   _put (cid, node, callback) {
     callback = callback || noop
 
     waterfall([
-      (cb) => this._getFormat(cid.codec, cb),
+      async () => {
+        return this._getFormat(cid.codec)
```

Comment:
The `async` module supports async/await style coding.

Reply:
Apparently this only works with non-transpiled async functions? "Using ES2017 async functions" -> https://caolan.github.io/async/ (can't deep link, sorry)

Reply:
Good catch. Does this mean that tests pass as we run webpack with different configurations for development and production? I even think we don't really run tests with the Browser build we publish.

Reply:
According to [Moz](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function#Browser_compatibility) async functions are supported everywhere except IE, and our contributing guidelines say [which Node version we support](https://github.com/ipfs/community/blob/master/CONTRIBUTING_JS.md#supported-versions) but not which browser types/versions.

Aegir uses [`babel/preset-env`](https://babeljs.io/docs/en/next/babel-preset-env.html) which I guess means it'll use native async unless you use IE, which will get a polyfill and this will break.

@alanshaw @olizilla @lidel - what browsers we officially support?

Reply:
For the browser build we currently use [this browser list](https://github.com/ipfs/aegir/blob/a5c1c32798579cc30e89f5c56089e8f0e16ef4df/src/utils.js#L30-L35):

[>1%, last 2 versions, Firefox ESR, not ie < 11](https://browserl.ist/?q=>1%25%2C+last+2+versions%2C+Firefox+ESR%2C+not+ie+<+11)

Which contains mobile browsers that don't support async/await.